pkg/services/dashboards/service/dashboard_service.go

```diff
@@ -1293,6 +1281,34 @@ func (dr *DashboardServiceImpl) FindDashboards(ctx context.Context, query *dashb
 	return dr.dashboardStore.FindDashboards(ctx, query)
 }
 
+func (dr *DashboardServiceImpl) fetchFolderNames(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery, response *dashboardv0alpha1.SearchResults) (map[string]string, error) {
+	folderIds := []string{}
+	folderSet := map[string]string{} // to avoid duplicates
+	for _, hit := range response.Hits {
+		if hit.Folder != "" && folderSet[hit.Folder] == "" {
+			folderIds = append(folderIds, hit.Folder)
+			folderSet[hit.Folder] = hit.Folder
+		}
+	}
+
+	// call this with elevated permissions so we can get folder names where user does not have access
+	// some dashboards are shared directly with user, but the folder is not accessible via the folder permissions
+	serviceCtx, serviceIdent := identity.WithServiceIdentity(ctx, query.OrgId)
+	folders, err := dr.folderService.GetFolders(serviceCtx, folder.GetFoldersQuery{
```

Comment:
IIUC, GetFolders through unistore will list ALL the folders ([here](https://github.com/grafana/grafana/blob/fde475e3d9a28b19fc6a1decd7fa911123c00a2e/pkg/services/folder/folderimpl/unifiedstore.go#L343-L348)) and then filter down the folders.

Reply:
Stepping back. Right now this is powered by storage. Should this be powered by search given we can already filter by Ids.

If not, we can implement a way to filter the results in storage (using a fake label selector, those support the `in` operator)

Reply:
This could be powered by search but this seems to be the source of truth. My concern ( maybe not valid ) is the search index could be out of sync in cases like someone moving a folder with lots of dashboards.

Reply:
Is listing all folders a huge hit?  If there were 1000 folders and we're querying 10,000 dashboards, I think we are better off getting all folders vs hitting an api 10,000 times.

Reply:
switched to use search ( we can easily swap implementations though now )